8312401: SymbolTable::do_add_if_needed hangs when called in InstanceKlass::add_initialization_error path with requesting length exceeds max_symbol_length

jianglizhou opened the pull request:
Please review the simple fix to resolve infinite loop in SymbolTable::do_add_if_needed caused by extra long symbol string that exceeds Symbol::max_length(). See JDK-8312401 for details.

The jtreg test is converted from a test case constructed by @cushon.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312401](https://bugs.openjdk.org/browse/JDK-8312401): SymbolTable::do_add_if_needed hangs when called in InstanceKlass::add_initialization_error path with requesting length exceeds max_symbol_length (**Bug** - P3)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14938/head:pull/14938` \
`$ git checkout pull/14938`

Update a local copy of the PR: \
`$ git checkout pull/14938` \
`$ git pull https://git.openjdk.org/jdk.git pull/14938/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14938`

View PR using the GUI difftool: \
`$ git pr show -t 14938`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14938.diff">https://git.openjdk.org/jdk/pull/14938.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14938#issuecomment-1642707549)

jianglizhou commented:
/contributor @cushon

dholmes-ora reviewed (commented):
I'm not sure silently ignoring this by returning null is the right thing to do. Shouldn't we just truncate the message in this case?

Or are we even doing the wrong thing by returning as a Symbol in the first place? By doing so we seem to impose an artificial VM limit on the exception message length.

jianglizhou commented:
> I'm not sure silently ignoring this by returning null is the right thing to do. Shouldn't we just truncate the message in this case?
> 
> Or are we even doing the wrong thing by returning as a Symbol in the first place? By doing so we seem to impose an artificial VM limit on the exception message length.

Max Symbol length is 64K. Exception message is bound to String length, 0x7FFFFFFF (2147483647).  I did some history search. Using Symbol for throwable message was initially done for ResolutionErrorTable::add_entry in JDK 8. That's been a long time. No issue has been surfaced for ResolutionError part. 

Currently java_lang_Throwable::detail_message can be called by the [following](https://github.com/search?q=repo%3Aopenjdk%2Fjdk%20java_lang_Throwable%3A%3Adetail_message%20&type=code):

- java_lang_Throwable::create_initialization_error
- Exceptions::log_exception
- ConstantPoolCache::save_and_throw_indy_exc
- exception_message
- add_resolution_error

No sure truncating the message would be better than returning a nullptr when the length exceeds max symbol length for above cases. Both behaviors are undefined by spec. Since these cases are VM internal implementation details only, maybe either truncating or nullptr is fine. Please let me know your preference.

I checked all callers and they can tolerate null message. For  java_lang_Throwable::create_initialization_error, when java_lang_Throwable::detail_message returns nullptr, it fills the new ExceptionInInitializerError with the thread and frame information.

I considered changing to use `char *` for the message in java_lang_Throwable::create_initialization_error (https://github.com/openjdk/jdk/pull/14938/commits/e2223f0613d53e89e11af587dc3d41a25ffd6849). But we need to handle all other callers as well. Exceptions::log_exception can use `char *` also. The other callers still require Symbol.

dholmes-ora commented:
> I considered changing to use char * for the message in java_lang_Throwable::create_initialization_error (https://github.com/openjdk/jdk/commit/e2223f0613d53e89e11af587dc3d41a25ffd6849). But we need to handle all other callers as well. Exceptions::log_exception can use char * also. The other callers still require Symbol.

Hmmm ... by using a Symbol we are artificially restricting the length of the exception message. That is fine if the VM is constructing the message, but not if the message can come from an external source as in this case. IIUC only `java_lang_Throwable::create_initialization_error` tries to use an existing, externally defined message, so only it should have this problem. So we either stop using a Symbol in this case (we copied the code from the resolution error case) or we handle the limitation - which to me means we should truncate so that we at least retain most of the potentially useful information. It seems unrealistic to have exception messages this large in non-tests. I'm not sure why we use Symbol here in the first place. @coleenp will know but is away this week.

jianglizhou commented:
> > I considered changing to use char * for the message in java_lang_Throwable::create_initialization_error ([e2223f0](https://github.com/openjdk/jdk/commit/e2223f0613d53e89e11af587dc3d41a25ffd6849)). But we need to handle all other callers as well. Exceptions::log_exception can use char * also. The other callers still require Symbol.
> 
> Hmmm ... by using a Symbol we are artificially restricting the length of the exception message. That is fine if the VM is constructing the message, but not if the message can come from an external source as in this case. IIUC only `java_lang_Throwable::create_initialization_error` tries to use an existing, externally defined message, so only it should have this problem. So we either stop using a Symbol in this case (we copied the code from the resolution error case) or we handle the limitation - which to me means we should truncate so that we at least retain most of the potentially useful information. 

Those are good points. I've updated the PR to use `char *` for message in java_lang_Throwable::create_initialization_error. 

I also changed java_lang_String::as_symbol and java_lang_String::as_symbol_or_null to check input string length and truncate to max symbol length if exceeding. With the fix in java_lang_Throwable::create_initialization_error, that would mostly be a no-op, but just to be safe. I also added asserts in SymbolTable::new_symbol to check length.

> It seems unrealistic to have exception messages this large in non-tests. I'm not sure why we use Symbol here in the first place. @coleenp will know but is away this week.

I'll wait for @coleenp's review/feedback.

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -569,9 +569,16 @@ char* java_lang_String::as_quoted_ascii(oop java_string) {
   return result;
 }
 
+// Return a Symbol for the input Java string. A new Symbol is created if one
+// does not exist. If the length of the input string exceeds Symbol::max_length,
+// the string is truncated to Symbol::max_length.
```
I don't think this is generally the correct way to approach this. We should only be using `String::as_symbol` when we know the String will fit into a Symbol - if it doesn't then we shouldn't be using this process. So any errors here should be fatal IMO. Each of the clients of this API should be examined to check valid usage i.e. it is okay for field/methods names due to 64K limit on those, but not for class/interface names - and certainly not for arbitrary strings like exception messages! Relatedly the usages of `java_lang_Throwable::detail_message` should also be examined.

jianglizhou replied:
> I don't think this is generally the correct way to approach this. We should only be using String::as_symbol when we know the String will fit into a Symbol - if it doesn't then we shouldn't be using this process. So any errors here should be fatal IMO. Each of the clients of this API should be examined to check valid usage i.e. it is okay for field/methods names due to 64K limit on those, but not for class/interface names - and certainly not for arbitrary strings like exception messages! Relatedly the usages of java_lang_Throwable::detail_message should also be examined.

All the other direct usages of java_lang_String::as_symbol are okay, except one from java_lang_Throwable::detail_message. All other callers convert a type string into a Symbol. 

I changed java_lang_String::as_symbol to truncate the string for java_lang_Throwable::detail_message usage mainly. Instead we should `guarantee` the string length to be < Symbol::max_length in  java_lang_String::as_symbol.

dholmes-ora reviewed (commented):
Okay on changing to `char*` for the exception message, but not sure about the rest. Will wait to see what @coleenp has to say. I think we have some dubious/fragile code here.

Thanks.

coleenp commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -2761,15 +2774,19 @@ Handle java_lang_Throwable::create_initialization_error(JavaThread* current, Han
+  oop detailed_message = java_lang_Throwable::message(throwable());
+  if (detailed_message != nullptr) {
+    message = java_lang_String::as_utf8_string(detailed_message);
+  }
```
It looks like you did remove converting the message to a Symbol here, so are there other places where long symbol names come through this code?  We should catch them at their sources.

jianglizhou replied:
> It looks like you did remove converting the message to a Symbol here, so are there other places where long symbol names come through this code? We should catch them at their sources.

Please see https://github.com/openjdk/jdk/pull/14938#issuecomment-1644534913. Exceptions::log_exception can be simply changed to use `char *`. Other callers might need a bit more work.

coleenp reviewed (commented):
I think just the javaClasses.cpp and the test should be enough to fix this bug.  Thank you for fixing it.

jianglizhou commented:
> I think just the javaClasses.cpp and the test should be enough to fix this bug. Thank you for fixing it.

Thanks for the review., @coleenp  I'll revert other changes. I'm still concerned about the usages of java_lang_Throwable::detail_message in other places, if we are not going to truncate the string length explicitly. I'll file a separate bug for investigating other callers for java_lang_Throwable::detail_message. @dholmes-ora also suggested.

jianglizhou commented:
> > I think just the javaClasses.cpp and the test should be enough to fix this bug. Thank you for fixing it.
> 
> Thanks for the review., @coleenp I'll revert other changes. I'm still concerned about the usages of java_lang_Throwable::detail_message in other places, if we are not going to truncate the string length explicitly. I'll file a separate bug for investigating other callers for java_lang_Throwable::detail_message. @dholmes-ora also suggested.

Filed https://bugs.openjdk.org/browse/JDK-8312616.

jianglizhou commented:
@coleenp @dholmes-ora Please take another look of the PR. I've reverted the java_lang_String::as_symbol change that truncates the string if length exceeds max symbol length, per our discussions in this thread. The PR now only addresses the issue with java_lang_Throwable::create_initialization_error. The other usages of java_lang_Throwable::detail_message can be followed up with https://bugs.openjdk.org/browse/JDK-8312616.

coleenp reviewed (approved):
Looks good.  Thank you for fixing this.

iklam commented on src/hotspot/share/classfile/symbolTable.cpp:
```diff
@@ -337,6 +337,7 @@ Symbol* SymbolTable::lookup_common(const char* name,
 }
 
 Symbol* SymbolTable::new_symbol(const char* name, int len) {
+  assert(len <= Symbol::max_length(), "sanity");
```
There are other calls to `java_lang_Throwable::detail_message()` that will overflow `Symbol::max_length()`, so adding this assert will just change the hang into a crash, neither is desirable.

If we want to properly fix `java_lang_Throwable::detail_message()` in a future RFE, it's better to temporarily change `java_lang_Throwable::detail_message()` in this PR to truncate the string (instead of hanging). That's a better alternative than crashing or hanging.

jianglizhou replied:
> There are other calls to `java_lang_Throwable::detail_message()` that will overflow `Symbol::max_length()`, so adding this assert will just change the hang into a crash, neither is desirable.
> 
> If we want to properly fix `java_lang_Throwable::detail_message()` in a future RFE, it's better to temporarily change `java_lang_Throwable::detail_message()` in this PR to truncate the string (instead of hanging). That's a better alternative than crashing or hanging.

Thanks for the thoughts. We have't seen issues in other call paths. It's better to fix the other callers properly with JDK-8312616 instead of using a workaround (truncate) with an undefined behavior.

Failing with assert provides more useful information than hang, however most end users only run on release binary (so most users would not see asserts). As no hang has been reported in other call paths, I think the new asserts would not be triggered in those cases with debug binaries. I can also revert the new asserts in this PR if there's a concern.

iklam replied:
OK, sounds good to me. I think the current PR is fine.

jianglizhou replied:
Clarified JBS issue title and PR title.

dholmes-ora reviewed (approved):
I agree with the changes in the PR as presented to fix the problem at hand. Any other misuses of String->Symbol conversion should be handled by the new RFE.

Thanks.

jianglizhou commented:
Thanks for the reviews and input.

jianglizhou commented:
/integrate

jianglizhou commented:
/backport jdk21u
